### PR TITLE
feat: provide vulnerability count as inline value on Open Source scan

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

_Provide description of this PR and changes, if linked Jira ticket doesn't cover it in full._

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: application/server/inline_values.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/snyk/snyk-ls/internal/uri"
+)
+
+func textDocumentInlineValueHandler(c *config.Config) jrpc2.Handler {
+	return handler.New(func(ctx context.Context, params lsp.InlineValueParams) (any, error) {
+		logger := c.Logger().With().Str("method", "textDocumentInlineValueHandler").Logger()
+		documentURI := params.TextDocument.URI
+		logger.Info().Msgf("Request for %s:%s RECEIVED", documentURI, params.Range.String())
+		defer logger.Info().Msgf("Request for %s:%s DONE", documentURI, params.Range.String())
+		if s, ok := di.Scanner().(snyk.InlineValueSupporter); ok {
```
Comment (PeterSchafer) [inline application/server/inline_values.go:39]: Question: is it correct that the range is not considered here at all? If so, maybe we don't need to print it as well.

File: application/server/inline_values.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+			filePath := uri.PathFromUri(documentURI)
+			values, err := s.GetInlineValues(filePath)
+			if err != nil {
+				return nil, err
+			}
+			lspInlineValues := converter.ToInlineValues(values)
```
Comment (PeterSchafer) [inline application/server/inline_values.go:45]: nitpick: it is a bit confusing that the output of `GetInlineValues()` needs to be converted by `ToInlineValues()`. So either of these method names is a bit misleading.

File: infrastructure/oss/vulnerability_count.go [hunk 3]
```diff
@@ -75,27 +85,71 @@ func (v *vulnerabilityCountImpl) ProcessVulnerabilityCount(
...
+		vci.severityCounts[snyk.High],
+		vci.severityCounts[snyk.Medium],
+		vci.severityCounts[snyk.Low],
+		vci.mostSevereVulnerabilityId,
+	)
+	return text
+}
+
+func (oss *Scanner) toInlineValueAndAddToCache(vci *VulnerabilityCountInformation) snyk.InlineValue {
+	values := oss.inlineValues[vci.FilePath]
```
Comment (PeterSchafer) [inline infrastructure/oss/vulnerability_count.go:134]: suggestion: it is a bit tricky to read this code as it uses values and value. maybe rename values to cachedValues, to better differentiate.

File: infrastructure/oss/vulnerability_count.go [hunk 4]
```diff
@@ -75,27 +85,71 @@ func (v *vulnerabilityCountImpl) ProcessVulnerabilityCount(
...
+func (oss *Scanner) addVulnerabilityCountsAsInlineValuesToCache(issues []snyk.Issue) {
+	c := config.CurrentConfig()
+	logger := c.Logger().With().Str("method", "addVulnerabilityCountsAsInlineValuesToCache").Logger()
+	vc := NewVulnerabilityCountService(c)
+	counts, err := vc.DetermineVulnerabilityCounts(issues)
+	if err != nil {
+		logger.Err(err).Msg("couldn't get vulnerability counts")
+		oss.errorReporter.CaptureError(err)
+	}
+	for _, r := range counts {
```
Comment (PeterSchafer) [inline infrastructure/oss/vulnerability_count.go:110]: suggestion: could we find better variable names for `r` and maybe even `vci`?

File: application/server/inline_values.go (hunk 1 above)
Comment (bastiandoetsch) [reply application/server/inline_values.go:39]: This may be incorrect, gotta check

File: application/server/inline_values.go (hunk 2 above)
Comment (bastiandoetsch) [reply application/server/inline_values.go:45]: It's consistent with the other converts from domain objects to lsp objects

File: infrastructure/oss/inline_value.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+package oss
+
+import (
+	"github.com/rs/zerolog/log"
+
+	"github.com/snyk/snyk-ls/domain/snyk"
+)
+
+func (oss *Scanner) GetInlineValues(path string, myRange snyk.Range) ([]snyk.InlineValue, error) {
```
Comment (PeterSchafer) [inline infrastructure/oss/inline_value.go:25]: nitpick: In this case not super important, but for structures it might be more efficient to use a pointers.

File: infrastructure/oss/inline_value.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	method := "GetInlineValues"
+	log.Debug().Str("method", method).Str("path", path).Msg("called")
+
+	value := oss.inlineValues[path][myRange.Start.Line]
```
Comment (PeterSchafer) [inline infrastructure/oss/inline_value.go:29]: Question: The current assumption here seems to me, that start line and end line are the same. Is this actually correct. In the lsp specification it says
> The document range for which inline values should be computed.

So I would be assuming that the range can span multiple lines and we would need to iterate over the inlineValues here to determine if they are in range or not.

Comment (bastiandoetsch) [reply infrastructure/oss/inline_value.go:29]: For OSS vulnerability count it's only the line that's relevant though.  That's why in general, we work with the range, but in the concrete implementation not.

Comment (PeterSchafer) [reply infrastructure/oss/inline_value.go:29]: I understand that the OSS vulnerability count is associated to a single line. But isn't this logic a lookup/search functionality that is triggered by the lsp client and the lsp client doesn't know anything about this, it only known the viewport of a document to show.

File: infrastructure/oss/inline_value.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+func (oss *Scanner) GetInlineValues(path string, myRange snyk.Range) (result []snyk.InlineValue, err error) {
+	logger := log.With().Str("method", "GetInlineValues").Logger()
+	logger.Debug().Str("path", path).Msg("called")
+
+	inlineValues := oss.inlineValues[path]
+	if inlineValues == nil {
+		logger.Debug().Str("path", path).Msg("no inlineValues found")
+		return nil, nil
```
Comment (PeterSchafer) [inline infrastructure/oss/inline_value.go:32]: Suggestion: Please return an empty slice here, otherwise [this](https://github.com/snyk/snyk-ls/pull/336/files#diff-a44b6d442dbb9afbfc2d35aa88806cfb9b0cb6c4a10508cb40e9c7404ec2fad6R83) will have an issue.

**Review by PeterSchafer — APPROVED** [review]

File: application/server/notification.go [hunk 8]
```diff
@@ -109,7 +109,6 @@ func registerNotifier(srv lsp.Server) {
 			if len(params.Diagnostics) > 0 {
 				source = params.Diagnostics[0].Source
 			}
-			handleCodelensRefresh(srv)
```
Comment (PeterSchafer) [inline application/server/notification.go:112]: Question: why was this removed?

Comment (PeterSchafer) [reply application/server/notification.go:112]: seems to be unnecessary since a test sends it anyways.

**Review by PeterSchafer — APPROVED** [review]
